zproject/backends.py

```diff
@@ -422,6 +422,12 @@ def authenticate(
                 raise JsonableError(_("You need to reset your password."))
 
         if is_password_correct:
+            if not check_password_strength(password):
+                if hasattr(request, "session"):
+                    request.session["needs_to_change_password"] = True
+                else:
+                    raise JsonableError(_("You need to reset your password."))
```

Comment:
I kinda feel like it'd be better for this to share its code path with `PasswordTooWeakError`; it's not worth having two different mechanisms for what happens when you're trying to login and your password is too weak.

I think we can just do
```
if is_password_correct and not check_password_strength(password):
     raise PasswordTooWeakError()
```

in the above block, and remove most of the other new code in this pull request.

The UX will be a little less friendly, in requiring a reset rather than letting the user change their password without the reset link flow, but it'll also be more secure, and avoid extra duplicated code paths here.

Reply:
Okay, but that will leave us unable to figure out if the `PasswordTooWeakError` was from the rehashing issue for us to be able to log that. Ended up using nested try, except blocks to be able to log that corner case.